### ISSUE TITLE
performance issue

### ISSUE THREAD

**Issue by hamed-musallam** [cheminfo/nmrium#1425]
@targos @lpatiny 

In some cases, we need to reset the state, in other words, create a new draft from the initial state, One of these cases when loading a JSON file.  To achieve that I use creatDraft to give me more flexibility and keep used of produce function for the other actions from immer.

This is the old way
```js
function innerReducerFunction(draft, action) {
  // reducer code goes here.
}
export const reducerFunction= produce(innerReducerFunction);

```

The new way 

```js
export function reducerFunction(state, action) {
   switch(action.type){
     case Load_JSON:{
         const draft = createDraft(initState);
      // handle something
          return finishDraft(draft );
     }
     default:
       return produce(state,(draft)=>{
         switch(action.type){
         // the rest of cases  
        }
        default:
      });
     }
  // reducer code goes here.
}

```

the problem that I face is the performance goes down when using the new way, is there something wrong in the new way?

**Comment by targos** [comment]
You do not need to return a draft from the reducer. You can keep the old way and in some cases return a plain new object.

**Comment by targos** [comment]
In other words: forget about createDraft/finishDraft. It should not be necessary to use them.

**Comment by hamed-musallam** [comment]
> You do not need to return a draft from the reducer. You can keep the old way and in some cases return a plain new object.

You mean I should use the new way but instead of using createDraft and finishDraft i can return new object

```
export function reducerFunction(state, action) {
   switch(action.type){
     case Load_JSON:{
       // handle something
          return {new state);
     }
     default:
       return produce(state,(draft)=>{
         switch(action.type){
         // the rest of cases  
        }
        default:
      });
     }
  // reducer code goes here.
}
```

**Comment by targos** [comment]
no, I mean you can return a new object from inside a produce callback (so keep the old way)

**Comment by hamed-musallam** [comment]
@targos 

I did this before and it give me this error 
[Immer] An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft.

**Comment by targos** [comment]
of course, if you return a new value, you should not modify the draft (only read from it)

**Comment by hamed-musallam** [comment]
I stopped all the code and return a new plain object !!!

**Comment by targos** [comment]
If you push a branch with the failing code, I can have a look

**Comment by hamed-musallam** [comment]
thank you @targos for your help, I resolve the problem, it was a mistake from me, I update a value in the draft at the beginning of the reducer function.

closed by fe64cb166b020b80d9665d4547cda30c3b8b69f7
